version: '3.9'

services:
  web:
    build:
      context: .
      dockerfile: Dockerfile
    environment:
      - PORT=8000
    ports:
      - "${FORWARD_WEB_PORT}:8000"
    volumes:
      - ${PWD}/:/opt/app/
      - site-packages:/usr/local/lib/python3.10/site-packages
    networks:
      - coinr
    depends_on:
      - postgres
      - redis

  postgres:
    image: postgres:14-alpine
    environment:
      - POSTGRES_USER=alaffia
      - POSTGRES_PASSWORD=${POSTGRES_PASSWORD}
      - POSTGRES_DB=coinr
    ports:
      - '${FORWARD_DATABASE_PORT}:5432'
    volumes:
      - 'postgres-data:/var/lib/postgresql/data'
    networks:
      - coinr

  redis:
    image: 'redis:alpine'
    ports:
      - '${FORWARD_REDIS_PORT}:6379'
    volumes:
      - 'redis-data:/data'
    networks:
      - coinr
    healthcheck:
      test: ["CMD", "redis-cli", "ping"]
      retries: 3
      timeout: 5s

  worker:
    build:
      context: ./
      dockerfile: Dockerfile
    command: >
      bash -c "
        cd coinr &&
        celery -A coinr worker -l INFO -Q coin-tasks --concurrency=1
      "
    volumes:
      - ${PWD}/coinr:/opt/app/coinr
    networks:
      - coinr
    depends_on:
      - redis


networks:
  coinr:
    driver: bridge

volumes:
  postgres-data:
  redis-data:
  celery-data:
  site-packages:
